### PR TITLE
Add wsypd, qsypd, ideal critical path, core hours and change asypd for psypd

### PR DESCRIPTION
**Pull request by paparici**
This pull request includes several changes to the `autosubmit_api` to add the calculation of the new metrics WSYPD, QSYPD, core-hours; the calculation of the critical path and its phases of an experiment. And, do a refactor to change the name of ASYPD to PSYPD. At some endpoints ASYPD and RSYPD are still there for retrocompatibility. Finally, a change to the warning message in case there are no POST jobs because the original warning said that the calculation would not be done because there are not POST jobs, but still was calculated. With the new changes, when there are no POST jobs; PSYPD = QSYPD so, the warning was changed. 

I would recommend at the future to change the definition of PSYPD to not let do the calculation in case there are no POST jobs. Therefore, change tests from the GUI and API. 

Issue #185

[CPMIP metrics issue ](https://github.com/BSC-ES/autosubmit-api/issues/188)

### CODE REVIEW

File: autosubmit_api/components/experiment/pkl_organizer.py [hunk 1]
```diff
@@ -107,11 +114,157 @@ def _sort_list_by_start_time(self, jobs: List[Job]):
...
+          
+    except Exception as exc:
+      self._add_warning(f"Could not assign parent-child relationships: {str(exc)}")
+
+  def find_ideal_critical_path(self) -> List[Job]:
+    """
+    Finds the critical path of jobs in the experiment.
+    The critical path is the path with the highest sum of queue time and run time.
+    Returns a list of jobs representing the critical path.
+    """
```
Comment (LuiggiTenorioK) [inline autosubmit_api/components/experiment/pkl_organizer.py:149]: The implementation here seems good. Anyway, I'll suggest this:

- [ ] Make a generalized function to find the critical path, receiving a dictionary (key, value). This is to separate the job logic from the algorithm.
- [ ] Make unit tests for the algorithm and also focus in testing edge cases (no jobs, more than one critical path, 0 run times, etc)

**Review by LuiggiTenorioK — CHANGES_REQUESTED** [review]
In general, the structure of the PR seems good and straightforward. Good job!

Anyway, I'm leaving some comments to focus on the consistency, robustness, and quality of the code. Take special attention to the documentation and testing parts :eyes:

File: autosubmit_api/performance/performance_metrics.py [hunk 2]
```diff
@@ -93,29 +100,44 @@ def _update_jobs_with_time_data(self):
...
+      self.valid_sim_core_hours_sum = self._calculate_sum_core_hours()
       self.SYPD = self._calculate_SYPD()
-      self.ASYPD = self._calculate_ASYPD()
+      self.PSYPD = self._calculate_PSYPD()
+      self.QSYPD = self._calculate_QSYPD()
       self.RSYPD = self._calculate_RSYPD()
       self.JPSY = self._calculate_JPSY()
       self.CHSY = self._calculate_CHSY()
+      self.ideal_critical_path = self.pkl_organizer.find_ideal_critical_path()
+      self.phases = self.pkl_organizer.calculate_ideal_critical_path_phases(self.ideal_critical_path)
```
[the remark this answers is not in the document]
Comment (MissterP) [reply autosubmit_api/performance/performance_metrics.py:119]: I added a try-catch to manage the exception in case there is no way to obtain the structure table.

File: docs/source/guides/performance/index.rst [hunk 3]
```diff
@@ -33,31 +33,37 @@ account:
...
-**Experiment's Real Simulated Years Per Day**: :math:`RSYPD = \frac{\sum_{i \in C} y_i \cdot 86400}{t_f - t_s}`
+WSYPD represents the effective throughput of the experiment's workflow. It is 
+calculated by multiplying the sum of simulated years from the valid 
+simulation jobs by the number of seconds in a day and dividing by the 
+total combined run and queue time measured along the workflow’s critical path.
 
-Generalization of SYPD and ASYPD
+.. important:: This metric was previously known as Real Simulated Years Per Day.
+
+**Experiment's Ideal Workflow Simulated Years Per Day**: :math:`IWSYPD = \frac{\sum_{i \in C} y_i \cdot 86400}{\sum_{i \in jobs critical path} (r_i)}}`
```
Comment (LuiggiTenorioK) [inline docs/source/guides/performance/index.rst:49]: image

Extra bracket at the end of the formula. Oops!

```suggestion
**Experiment's Ideal Workflow Simulated Years Per Day**: :math:`IWSYPD = \frac{\sum_{i \in C} y_i \cdot 86400}{\sum_{i \in jobs critical path} (r_i)}`
```

File: docs/source/guides/performance/index.rst [hunk 4]
```diff
@@ -33,31 +33,37 @@ account:
...
-The computation of *Real Simulated Year Per Day* will only consider experiments
-that have a TRANSFER or a CLEAN job at the end of the workflow. AutosubmitAPI
-checks when the first SIM job started (not submitted), and when the
-last TRANSFER job finished. If the workflow does not have a TRANSFER job, then it
-checks the last CLEAN job finish time.
+.. important:: This metric was previously known as Actual Simulated Years Per Day.
 
-Let :math:`t_s` be the time, in seconds, that the first SIM job started. Let
-:math:`t_f` be the time in which the last TRANSFER or CLEAN job ended.
+**Experiment's Workflow Simulated Years Per Day**: :math:`WSYPD = \frac{\sum_{i \in C} y_i \cdot 86400}{\sum_{i \in jobs critical path} (q_i + r_i)}`
```
Comment (LuiggiTenorioK) [inline docs/source/guides/performance/index.rst:40]: image

`jobs critical path` is hard to read in the formatted formula. A better approach is to assign that to a math variable like here:

image

You can add something like: "Let P be the critical path ..."
